File: backend/app/schemas/reports.py
from datetime import datetime

from pydantic import BaseModel, Field


class ReportGenerate(BaseModel):
    student_id: int
    audience: str = Field(pattern="^(student|tutor|parent)$")
    subject_id: int | None = None


class ReportOut(BaseModel):
    id: int
    student_id: int
    subject_id: int | None
    audience: str
    status: str
    title: str
    created_at: datetime
    generated_at: datetime | None


class ReportDetail(ReportOut):
    content: str | None
    error: str | None

File: frontend/src/api/reports.ts
import type { components } from "./schema";
import { api } from "./client";

/** `audience` and `status` are plain `str` in the OpenAPI schema; the literal
    unions are frontend refinements (surfaces branch on the exact value), kept
    here while the rest of the shape is anchored to the generated type. */
export type Report = Omit<components["schemas"]["ReportOut"], "audience" | "status"> & {
  audience: "student" | "tutor" | "parent";
  status: "generating" | "ready" | "failed";
};

export type ReportDetail = Omit<components["schemas"]["ReportDetail"], "audience" | "status"> & {
  audience: "student" | "tutor" | "parent";
  status: "generating" | "ready" | "failed";
};

export const listReports = (studentId: number) =>
  api<Report[]>(`/api/v1/reports?student_id=${studentId}`);
export const getReport = (id: number) => api<ReportDetail>(`/api/v1/reports/${id}`);
export const generateReport = (payload: {
  student_id: number;
  audience: string;
  subject_id?: number | null;
}) =>
  api<ReportDetail>("/api/v1/reports/generate", {
    method: "POST",
    body: JSON.stringify(payload),
  });
